src/python/pants/core/goals/test.py

```diff
@@ -242,8 +257,8 @@ def materialize(self, console: Console, workspace: Workspace) -> Tuple[PurePath,
                 report_paths.append(report_path)
         return tuple(report_paths)
 
-    def artifacts(self) -> Optional[Dict[str, Snapshot]]:
-        artifacts = {}
+    def artifacts(self) -> Optional[Dict[str, Union[Snapshot, FileDigest]]]:
+        artifacts: Dict[str, Union[Snapshot, FileDigest]] = {}
```

Comment:
FYI, you can leverage the new type hints syntax in several places. Totally fine to not do this to cut down on the diff, though.

Here, it would be:

```python
artifacts: dict[str, Snapshot | FileDigest]
```